### PR TITLE
Spe refactor and advection fix

### PR DESCRIPTION
**Pull request by Chris-Pedersen**
I wanted to make sure our calculation of spectral energy transfer is correct, considering a) this is our main target metric, and b) arakawa+smag models appear to be anti-diffusive even on small scales, which is counter intuitive. [From eq E2 of Pavel's paper](https://arxiv.org/pdf/2302.07984.pdf), this is just the cross-spectrum of streamfunction and whatever rhs is being applied. Pyqg uses a more complicated decomposition into ape, and keflux, but we do not really care about this. So both as a sanity/consistency check, and also just to simplify the code, lets implement this new, more simple calculation.

### CODE REVIEW

**Comment by Chris-Pedersen** [comment]
These two implementations are consistent. So will delete the ape+kef one (we can always reinstate it since we have the syntax)


Screenshot 2023-12-14 at 6 49 53 PM


Screenshot 2023-12-14 at 6 45 10 PM

**Comment by Chris-Pedersen** [comment]
Some discrepency between spectral energy transfers for 256**2 and 64**2 sims, with same config and both in pseudospectral kernel. Question is does this come from the simulation/kernel (which should be the same), or the spectral energy transfer calculation (which should also be the same).

Screenshot 2023-12-14 at 8 25 56 PM
Screenshot 2023-12-14 at 8 25 50 PM

**Comment by Chris-Pedersen** [comment]
KE spectra is a simple calculation, looks like the issue is in the fields itself. These figures are relevant for #4 

Screenshot 2023-12-14 at 8 32 48 PM

**Comment by Chris-Pedersen** [comment]
Comparing these 2 ways of calculating spectral energy transfer, they don't agree:

Screenshot 2023-12-19 at 12 50 27 PM

check doing the same in pyqg

**Comment by Chris-Pedersen** [comment]
Ok I'll come back to this at a later date. I mostly wanted to verify that our spectral energy transfer calculation was correct/consistent with pyqg. We show that in the following:

```
def plot_qs():
    fig, axs = plt.subplots(2, 3,figsize=(12,6))
    plt.suptitle("Fields after %d timesteps after syncing" % timestep)
    axs[0,0].set_title("pyqg fields")
    im1=axs[0,0].imshow(m_64.q[0])
    im2=axs[1,0].imshow(m_64.q[1])

    plt.colorbar(im1, ax=axs[0,0])
    plt.colorbar(im2, ax=axs[1,0])

    axs[0,1].set_title("torchqg fields")
    im3=axs[0,1].imshow(torchqg_64.q[0])
    im4=axs[1,1].imshow(torchqg_64.q[1])

    plt.colorbar(im3, ax=axs[0,1])
    plt.colorbar(im4, ax=axs[1,1])

    axs[0,2].set_title("pyqg-torch")
    im9=axs[0,2].imshow(m_64.q[0]-torchqg_64.q[0].numpy())
    im10=axs[1,2].imshow(m_64.q[1]-torchqg_64.q[1].numpy())

    plt.colorbar(im9, ax=axs[0,2])
    plt.colorbar(im10, ax=axs[1,2])

    plt.tight_layout()

YEAR = 24*60*60*360.
m_64=pyqg.QGModel(nx=64,dt=3600,log_level=0,tavestart=5*YEAR)
## Run a pyqg sim
sampling_freq=1000

snapshots = []
for aa in tqdm(range(int(8e4))):
    if m_64.tc % sampling_freq == 0:
        ds_64=m_64.to_dataset().copy(deep=True)
        snapshots.append(ds_64)
    m_64._step_forward()


timestep=0
torchqg_64=torch_model.PseudoSpectralModel(nx=64,dt=3600)
torchqg_64.set_q1q2(torch.tensor(m_64.q))


for aa in range(5000):
    torchqg_64._step_ab3()
    m_64._step_forward()
    m_64._calc_derived_fields()
    m_64._invert()
    m_64._do_advection()
    m_64._do_friction()
    m_64._do_external_forcing()
    m_64._calc_diagnostics()
    torchqg_64._calc_derived_fields()
    timestep+=1


plot_qs()


kef_qg=(np.real(m_64.del1*m_64.ph[0]*np.conj(m_64.Jpxi[0])) + np.real(m_64.del2*m_64.ph[1]*np.conj(m_64.Jpxi[1])))/m_64.M**2
ape_qg=(m_64.rd**-2 * m_64.del1*m_64.del2 * np.real((m_64.ph[0]-m_64.ph[1])*np.conj(m_64.Jptpc)))/m_64.M**2

kef_t=((np.real(torchqg_64.del1*torchqg_64.ph[0]*np.conj(torchqg_64.Jpxi[0])) + np.real(torchqg_64.del2*torchqg_64.ph[1]*np.conj(torchqg_64.Jpxi[1])))/torchqg_64.M**2)
ape_t=(torchqg_64.rd**-2*torchqg_64.del1*torchqg_64.del2*np.real((torchqg_64.ph[0]-torchqg_64.ph[1])*np.conj(torchqg_64.Jptpc))/torchqg_64.M**2)

spe_qg=kef_qg+ape_qg
spe_t=kef_t+ape_qg

## ispec from torchqg
plt.title("Spectral energy transfer %d timesteps after syncing fields" % timestep)
plt.semilogx(torchqg_64.get_ispec_1(spe_qg),label="pyqg")
plt.semilogx(torchqg_64.get_ispec_1(spe_t),label="torch qg",linestyle="dashed")
plt.legend()
plt.tight_layout()
```

which produces

Screenshot 2023-12-20 at 2 31 42 PM

Screenshot 2023-12-20 at 2 32 03 PM


so from this we can conclude:
1. Pyqg and torchqg psuedospectral numerical solvers are consistent close to numerical precision
2. The spectral energy transfer calculations are consistent

**Comment by Chris-Pedersen** [comment]
Still work to do here but we are at a sensible stopping point, so merging now. The spectra now agree for pyqg and torchqg - the issue was [here](https://github.com/Chris-Pedersen/torch_qg/pull/29/commits/76e69450fa3fa18ff81744a604fc8d6d3e8ec461#diff-7e019166fbeb874e215cb96ff9bf7d4cb16dc79ebfa428410b6e9803707abc1aR153), in that we were using the _plot_ bins from pyqg to do the radial averaging, which are shifted by dk/2.


pyqgtorchqg (1)

Arakawa kernel still has confusingly godo performance even for super low res, and the two approaches to calculate spectral energy transfer don't agree, but will address these in future PRs.
Screenshot 2024-01-12 at 11 36 52 AM
